lib/tpm2_alg_util.c

```diff
@@ -120,3 +122,123 @@ TPM_ALG_ID tpm2_alg_util_from_optarg(char *optarg) {
+    case TPM_ALG_SHA512:
+        /* fallsthrough */
+    case TPM_ALG_SM3_256:
+        return true;
+        /* no default */
+    }
+
+    return false;
+}
+
+static const char *hex_to_byte_err (int rc) {
+    switch (rc) {
+    case -2:
+        return "String not even in length";
+    case -3:
+        return "Non hex digit found";
+    case -4:
+        return "Hex value too big for digest";
+    }
+    return "unknown";
+}
+
+/* parse digest=<hash|file> digest=<hash|file> ... */
+bool pcr_parse_digest_list(char *digest_list, TPML_DIGEST_VALUES *digests) {
+
+    /* XXX Should we ensure initialization or force the caller? */
+    digests->count = 0;
+
+    char *tok;
+    char *save_ptr = NULL;
+    UINT32 count = 0;
+    while( (tok = strtok_r(digest_list, " ", &save_ptr))) {
+        digest_list = NULL;
+
+        if (count > ARRAY_LEN(digests->digests)) {
+            LOG_ERR("Specified too many digests, max is: %zu",
+                    ARRAY_LEN(digests->digests));
+            return false;
+        }
+
+        TPMT_HA *d = &digests->digests[count];
+
+        /*
+         * Split <hash>:<data> on : and seperate with null byte, ie:
+         * <hash>'\0'<data>
+         */
+        char *split = strchr(tok, ':');
+        *split = '\0';
+
+        char *data = split + 1;
+        char *stralg = tok;
+
+        /*
+         * Convert and validate the hash algorithm. It should be a hash algorithm
+         */
+        TPM_ALG_ID alg = tpm2_alg_util_from_optarg(stralg);
+        if (alg == TPM_ALG_ERROR) {
+            LOG_ERR("Could not convert algorithm, got: \%s\"", stralg);
+            return false;
+        }
+
+        bool is_hash_alg = tpm2_alg_util_is_hash_alg(alg);
+        if (!is_hash_alg) {
+            LOG_ERR("Algorithm is not a hash algorithm, got: \%s\"", stralg);
+            return false;
+        }
+
+        /* fill up the TPMT_HA structure with algorithm and digest */
+        BYTE *digest_data = (BYTE *)&d->digest;
+
+        /* TODO CHECK FOR DUPLICATE SPECIFICATIONS */
+        d->hashAlg = alg;
+
+        /*
+         * always attempt to open it first, if it fails, then try to convert it as raw data.
+         * This allows users to have hex named files, and a file will always override the string.
+         * Files can be named anything that doesn't have a space.
+         */
+        UINT16 size = sizeof(d->digest);
+        FILE *f = fopen(data, "rb");
+        if (f) {
+            bool result = files_load_bytes_from_stream(f, digest_data, &size);
+            if (!result) {
+                fclose(f);
+                return false;
+            }
+            fclose(f);
+        } else {
+            /* strip any preceding hex on the data as tpm2_util_hex_to_byte_structure doesn't support it */
+            bool is_hex = !strncmp("0x", data, 2);
+            if (is_hex) {
+                data += 2;
+            }
+
+            int rc = tpm2_util_hex_to_byte_structure(data, &size, digest_data);
```

Comment:
This routine sucks... can't handle odd sizes by assuming 0 starting nibble, can't handle 0x prefix...